### PR TITLE
feat: TreeSelectField

### PR DESCRIPTION
**Pull request by bdow**
Creates new TreeSelectField component that allows nested groupings of items within the menu. Handles updating parent and child options selected state based on user input or values defined. Includes logic for lazily loading options.
Supports all existing states as Beam's other fields (disabled, readOnly, etc), as well as variations (compact, label styles, etc).

Adds tests and stories

Future improvements:
- Generate unique keys for each option based on grouping hierarchy
- Create Bound___Field equivalent
- Create Tree filter field for use within Beam filters
- Look at either moving existing TreeFieldState into either the context that currently only supports the collapsed state, or moving everything into a mobx state. Admittedly, this optimization should probably happen sooner than later. But looking to get something out in the meantime.

### CODE REVIEW

File: src/inputs/TreeSelectField/TreeSelectField.tsx [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+    // where we might want to do flatmap and return diff kind of array (children ? add level prop) inside children callback - can put markup wrapper div adds padding
+    // so we're not doing it multiple places
+    items: fieldState.filteredOptions,
+    isDisabled,
+    isReadOnly,
+    onInputChange,
+    onOpenChange,
+    children: ([item]: LeveledOption<O>) => (
+      // what we're telling it to render. look at padding here - dont have to pass down to tree option - filtered options is where we're flat mapping
+      <Item key={valueToKey(getOptionValue(item))} textValue={getOptionLabel(item)}>
```
Comment (blambillotte) [inline src/inputs/TreeSelectField/TreeSelectField.tsx:303]: I don't remember where this landed, are we just trusting for now that values are tagged IDs and will always be unique?

File: src/inputs/TreeSelectField/TreeSelectField.tsx [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+          leaf: { values: leafValues, options: leafOptions },
+          root: { values: rootValues, options: rootOptions },
+        });
+      }
+    },
+  });
+
+  function resetField() {
+    const { inputValue, selectedOptions } = fieldState;
+    if (inputValue !== "" || (selectedOptions.length === 1 && inputValue !== getOptionLabel(selectedOptions[0]))) {
```
Comment (blambillotte) [inline src/inputs/TreeSelectField/TreeSelectField.tsx:436]: maybe an an in-source comment here explaining what `reset` actually means? I'm not quite following the logic for this `inputValue` - I naively expect it to just clear on a `reset` type action?

**Review by blambillotte — APPROVED** [review]
Nice work! Excited to start using this component!

File: src/inputs/TreeSelectField/TreeSelectField.tsx [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+        }));
+      }
+    },
+    [],
+  );
+
+  // Only on the first open of the listbox, we want to load the options if they haven't been loaded yet.
+  const firstOpen = useRef(true);
+  function onOpenChange(isOpen: boolean) {
+    if (!isOpen) debugger;
```
[the remark this answers is not in the document]
Comment (bdow) [reply src/inputs/TreeSelectField/TreeSelectField.tsx:281]: oh jeesh. No good. I guess this wasn't nearly as ready for primetime as I thought. 🙈

File: src/inputs/TreeSelectField/TreeSelectField.tsx (hunk 1 above)
Comment (bdow) [reply src/inputs/TreeSelectField/TreeSelectField.tsx:303]: For now. I was going to have it be a follow up.

File: src/inputs/TreeSelectField/TreeOption.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  const { collapsedKeys, setCollapsedKeys, getOptionValue } = useTreeSelectFieldProvider<O, Value>();
+
+  const { optionProps, isDisabled, isFocused, isSelected } = useOption(
+    { key: item.key, shouldSelectOnPressUp: true, shouldFocusOnHover: false },
+    state,
+    ref,
+  );
+
+  // If this item is not selected, then determine if some of its children are selected to show the indeterminate state.
+  const isIndeterminate = !isSelected && option.children?.some((o) => hasSelectedChildren(o, state, getOptionValue));
```
[the remark this answers is not in the document]
Comment (bdow) [reply src/inputs/TreeSelectField/TreeOption.tsx:39]: That's right. I will note this in the comment here.

File: src/inputs/TreeSelectField/TreeOption.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  item: Node<LeveledOption<O>>;
+  state: ListState<O>;
+  contrast?: boolean;
+  allowCollapsing?: boolean;
+}
+/** Represents a single option within a ListBox - used by SelectField, MultiSelectField, and TreeSelectField */
+export function TreeOption<O>(props: TreeOptionProps<O>) {
+  const { item, state, contrast = false, allowCollapsing = true } = props;
+  const leveledOption = item.value;
+  if (!leveledOption) return null;
```
[the remark this answers is not in the document]
Comment (bdow) [reply src/inputs/TreeSelectField/TreeOption.tsx:23]: It shouldn't be. `value` will always be a `LeveledOption`, which is a tuple of `[NestedOption, number]`

File: src/inputs/TreeSelectField/TreeOption.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                e.stopPropagation();
+                setCollapsedKeys((prevKeys) =>
+                  collapsedKeys.includes(item.key) ? prevKeys.filter((k) => k !== item.key) : [...prevKeys, item.key],
+                );
+                return false;
+              }}
+              css={Css.br4.hPx(16).wPx(16).bgTransparent.onHover.bgGray300.$}
+              {...tid[`collapseToggle_${item.key}`]}
+            >
+              <Icon icon={collapsedKeys.includes(item.key) ? "triangleRight" : "triangleDown"} inc={2} />
```
[the remark this answers is not in the document]
Comment (bdow) [reply src/inputs/TreeSelectField/TreeOption.tsx:74]: Could be cool to do one day, for sure.

File: src/inputs/TreeSelectField/TreeSelectField.tsx [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+
+export interface TreeSelectFieldProps<O, V extends Value> extends BeamFocusableProps, PresentationFieldProps {
+  /** Renders `opt` in the dropdown menu, defaults to the `getOptionLabel` prop. `isUnsetOpt` is only defined for single SelectField */
+  getOptionMenuLabel?: (opt: O, isUnsetOpt?: boolean) => string | ReactNode;
+  getOptionValue: (opt: O) => V;
+  getOptionLabel: (opt: O) => string;
+  /** The current value; it can be `undefined`, even if `V` cannot be. */
+  values: V[] | undefined;
+  onSelect: (options: TreeSelectResponse<O, V>) => void;
+  // disabledOptions?: (V | { value: V; reason: string })[];
```
[the remark this answers is not in the document]
Comment (bdow) [reply src/inputs/TreeSelectField/TreeSelectField.tsx:43]: Ahh, no. I meant to just remove. We can always add later if needed.

**Pull request merged**